File: apps/admin-v2/src/infrastructure/api/repositories/api-user.repository.ts
import type { CreateMemberInput, UserRepository } from '@/domain/repositories/user.repository';
import type { User, UserRole } from '@/domain/entities/user';
import type { PaginatedResult } from '@/shared/types/api';
import api from '../client';
import { mapUser } from '../mappers/user.mapper';
import { mapPaginatedResponse } from '../mappers/pagination';

export class ApiUserRepository implements UserRepository {
  async getAll(filters?: { role?: UserRole; excludeRole?: UserRole }): Promise<PaginatedResult<User>> {
    const params: Record<string, unknown> = {};
    if (filters?.role) params.role = filters.role;
    if (filters?.excludeRole) params.exclude_role = filters.excludeRole;

    const { data: res } = await api.get('/users', { params });
    return mapPaginatedResponse(res, mapUser);
  }

  async getById(id: string): Promise<User> {
    const { data: res } = await api.get(`/users/${id}`);
    return mapUser(res.data);
  }

  async invite(input: CreateMemberInput): Promise<User> {
    const payload = {
      name: input.name,
      username: input.username,
      password: input.password,
      email: input.email || undefined,
      phone: input.phone || undefined,
      role: input.role,
    };
    const { data: res } = await api.post('/users/invite', payload);
    return mapUser(res.data.user ?? res.data);
  }

  async changeRole(id: string, role: UserRole): Promise<User> {
    const { data: res } = await api.patch(`/users/${id}/role`, { role });
    return mapUser(res.data);
  }

  async resetPassword(id: string, password: string): Promise<void> {
    await api.patch(`/users/${id}/password`, { password });
  }
}

File: apps/admin_mobile/lib/infrastructure/api/repositories/api_user_repository.dart
import '../../../domain/entities/user.dart';
import '../../../domain/repositories/user_repository.dart';
import '../../../shared/types/paginated_result.dart';
import '../dio_client.dart';
import '../mappers/user_mapper.dart';
import 'paginated_helper.dart';

class ApiUserRepository implements UserRepository {
  final DioClient _client;

  ApiUserRepository(this._client);

  @override
  Future<PaginatedResult<User>> getAll(
      {UserRole? role, UserRole? excludeRole}) async {
    final params = <String, dynamic>{};
    if (role != null) params['role'] = role.apiValue;
    if (excludeRole != null) params['exclude_role'] = excludeRole.apiValue;

    final response = await _client.dio.get('/users', queryParameters: params);
    return extractPaginated(response.data, mapUser);
  }

  @override
  Future<User> getById(int id) async {
    final response = await _client.dio.get('/users/$id');
    final data = response.data['data'] ?? response.data;
    return mapUser(data as Map<String, dynamic>);
  }

  @override
  Future<User> invite(String email, UserRole role) async {
    final response = await _client.dio.post('/users/invite', data: {
      'email': email,
      'role': role.apiValue,
    });
    final data = response.data['data'] ?? response.data;
    return mapUser(data as Map<String, dynamic>);
  }

  @override
  Future<User> changeRole(int id, UserRole role) async {
    final response = await _client.dio.patch('/users/$id/role', data: {
      'role': role.apiValue,
    });
    final data = response.data['data'] ?? response.data;
    return mapUser(data as Map<String, dynamic>);
  }
}
